File: sdks/python/octollm-sdk/octollm_sdk/exceptions.py
"""
OctoLLM SDK Exception Classes

Custom exceptions for clear error handling and debugging.
"""

from typing import Any


class OctoLLMError(Exception):
    """Base exception for all OctoLLM SDK errors."""

    def __init__(
        self,
        message: str,
        status_code: int | None = None,
        details: dict[str, Any] | None = None,
        request_id: str | None = None,
    ):
        super().__init__(message)
        self.message = message
        self.status_code = status_code
        self.details = details or {}
        self.request_id = request_id

    def __str__(self) -> str:
        base = f"{self.__class__.__name__}: {self.message}"
        if self.status_code:
            base += f" (HTTP {self.status_code})"
        if self.request_id:
            base += f" [Request ID: {self.request_id}]"
        return base


class AuthenticationError(OctoLLMError):
    """Raised when authentication fails (401)."""

    def __init__(self, message: str = "Authentication failed", **kwargs: Any):
        kwargs.setdefault("status_code", 401)
        super().__init__(message, **kwargs)


class AuthorizationError(OctoLLMError):
    """Raised when authorization fails (403)."""

    def __init__(self, message: str = "Insufficient permissions", **kwargs: Any):
        kwargs.setdefault("status_code", 403)
        super().__init__(message, **kwargs)


class ValidationError(OctoLLMError):
    """Raised when request validation fails (400, 422)."""

    def __init__(self, message: str = "Validation error", **kwargs: Any):
        kwargs.setdefault("status_code", 422)
        super().__init__(message, **kwargs)


class NotFoundError(OctoLLMError):
    """Raised when a resource is not found (404)."""

    def __init__(self, message: str = "Resource not found", **kwargs: Any):
        kwargs.setdefault("status_code", 404)
        super().__init__(message, **kwargs)


class RateLimitError(OctoLLMError):
    """Raised when rate limit is exceeded (429)."""

    def __init__(
        self,
        message: str = "Rate limit exceeded",
        retry_after: int | None = None,
        **kwargs: Any,
    ):
        kwargs.setdefault("status_code", 429)
        super().__init__(message, **kwargs)
        self.retry_after = retry_after

    def __str__(self) -> str:
        base = super().__str__()
        if self.retry_after:
            base += f" (Retry after {self.retry_after}s)"
        return base


class ServiceUnavailableError(OctoLLMError):
    """Raised when a service is unavailable (503)."""

    def __init__(self, message: str = "Service temporarily unavailable", **kwargs: Any):
        kwargs.setdefault("status_code", 503)
        super().__init__(message, **kwargs)


class TimeoutError(OctoLLMError):
    """Raised when a request times out."""

    def __init__(self, message: str = "Request timeout", **kwargs: Any):
        kwargs.setdefault("status_code", 408)
        super().__init__(message, **kwargs)


class APIError(OctoLLMError):
    """Raised for general API errors (500+)."""

    def __init__(self, message: str = "API error occurred", **kwargs: Any):
        super().__init__(message, **kwargs)

File: sdks/typescript/octollm-sdk/src/exceptions.ts
/**
 * Custom exception classes for OctoLLM SDK
 */

export interface ErrorResponse {
  error: {
    code: string;
    message: string;
    details?: Record<string, any>;
  };
  request_id?: string;
}

/**
 * Base exception for all OctoLLM errors
 */
export class OctoLLMError extends Error {
  public readonly statusCode?: number;
  public readonly errorCode?: string;
  public readonly details?: Record<string, any>;
  public readonly requestId?: string;

  constructor(message: string, errorData?: ErrorResponse, statusCode?: number) {
    super(message);
    this.name = this.constructor.name;
    this.statusCode = statusCode;

    if (errorData) {
      this.errorCode = errorData.error.code;
      this.details = errorData.error.details;
      this.requestId = errorData.request_id;
    }

    Error.captureStackTrace(this, this.constructor);
  }
}

/**
 * Authentication failed (401)
 */
export class AuthenticationError extends OctoLLMError {
  constructor(errorData: ErrorResponse) {
    super(errorData.error.message || 'Authentication failed', errorData, 401);
  }
}

/**
 * Authorization failed (403)
 */
export class AuthorizationError extends OctoLLMError {
  constructor(errorData: ErrorResponse) {
    super(errorData.error.message || 'Authorization failed', errorData, 403);
  }
}

/**
 * Validation failed (400, 422)
 */
export class ValidationError extends OctoLLMError {
  constructor(errorData: ErrorResponse) {
    super(errorData.error.message || 'Validation failed', errorData, 400);
  }
}

/**
 * Resource not found (404)
 */
export class NotFoundError extends OctoLLMError {
  constructor(errorData: ErrorResponse) {
    super(errorData.error.message || 'Resource not found', errorData, 404);
  }
}

/**
 * Rate limit exceeded (429)
 */
export class RateLimitError extends OctoLLMError {
  public readonly retryAfter?: number;

  constructor(errorData: ErrorResponse, retryAfter?: number) {
    super(errorData.error.message || 'Rate limit exceeded', errorData, 429);
    this.retryAfter = retryAfter;
  }
}

/**
 * Service unavailable (503)
 */
export class ServiceUnavailableError extends OctoLLMError {
  constructor(errorData: ErrorResponse) {
    super(errorData.error.message || 'Service unavailable', errorData, 503);
  }
}

/**
 * Request timeout
 */
export class TimeoutError extends OctoLLMError {
  constructor(message: string) {
    super(message, undefined, 408);
  }
}

/**
 * Generic API error
 */
export class APIError extends OctoLLMError {
  constructor(errorData: ErrorResponse, statusCode?: number) {
    super(errorData.error.message || 'API error occurred', errorData, statusCode);
  }
}
